pkg/openthread/contrib/platform_alarm.c

```diff
@@ -42,31 +60,64 @@ void otPlatAlarmMilliStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
 {
     (void)aInstance;
     (void)aT0;
+    DEBUG("openthread: otPlatAlarmMilliStartAt: aT0: %" PRIu32 ", aDT: %" PRIu32 "\n", aT0, aDt);
 
-    DEBUG("openthread: otPlatAlarmStartAt: aT0: %" PRIu32 ", aDT: %" PRIu32 "\n", aT0, aDt);
-    ot_alarm_msg.type = OPENTHREAD_XTIMER_MSG_TYPE_EVENT;
-
-    if (aDt == 0) {
-        msg_send(&ot_alarm_msg, thread_getpid());
-    }
-    else {
-        int dt = aDt * US_PER_MS;
-        xtimer_set_msg(&ot_timer, dt, &ot_alarm_msg, thread_getpid());
-    }
+    uint32_t dt = aDt * US_PER_MS;
+    xtimer_set(&ot_millitimer, dt);
 }
 
 /* OpenThread will call this to stop alarms */
 void otPlatAlarmMilliStop(otInstance *aInstance)
 {
     (void)aInstance;
-    DEBUG("openthread: otPlatAlarmStop\n");
-    xtimer_remove(&ot_timer);
+    DEBUG("openthread: otPlatAlarmMilliStop\n");
+    xtimer_remove(&ot_millitimer);
 }
 
 /* OpenThread will call this for getting running time in millisecs */
 uint32_t otPlatAlarmMilliGetNow(void)
 {
-    uint32_t now = xtimer_now_usec() / US_PER_MS;
-    DEBUG("openthread: otPlatAlarmGetNow: %" PRIu32 "\n", now);
+    uint32_t now = (uint32_t)(xtimer_now_usec64() / US_PER_MS);
+    DEBUG("openthread: otPlatAlarmMilliGetNow: %" PRIu32 "\n", now);
+    return now;
+}
+
+/* Interupt handler for OpenThread micro-timer event */
+static void _microtimer_cb(void* arg) {
+    (void)arg;
+    msg_t microtimer_msg;
+    microtimer_msg.type = OPENTHREAD_MICROTIMER_MSG_TYPE_EVENT;
+    if (msg_send(&microtimer_msg, openthread_get_pid()) <= 0) {
+        DEBUG("openthread: possibly lost timer interrupt.\n");
+    }
+}
+
+/**
+ * Set the alarm to fire at @p aDt microseconds after @p aT0.
+ *
+ * @param[in] aInstance  The OpenThread instance structure.
+ * @param[in] aT0        The reference time.
+ * @param[in] aDt        The time delay in microseconds from @p aT0.
+ */
+void otPlatAlarmMicroStartAt(otInstance *aInstance, uint32_t aT0, uint32_t aDt)
```

Comment:
I have never worked with openthread so far, so please forgive if I'm asking nonsense.

The implementation in this PR would allow one timer in total to be set with `otPlatAlamMicroStartAt()` and one with the milliseconds version. (And the version in master would allow one in total.) This seems to be strange to me.

If I had to guess, I would say one timer per `otInstance` would make most sense. And one timer in total would make second-most sense to. Can someone with the required knowledge comment on this?

I think it should be possible to only provide one common internal implementation and offer `otPlatAlarmMicroStartAt()` and `otPlatAlarmMilliStartAt()` as `static inline` wrappers on top if this in order to safe some ROM.

Reply:
We need to call otPlatAlarmMilliFired(sInstance) when a millitimer is fired and otPlatAlarmMicroFired(sInstance) when a microtimer is fired. Without separation, I think it is hard to know what timer is fired. FYI, https://github.com/openthread/openthread/blob/master/src/core/common/timer.cpp